src/middlewared/middlewared/plugins/smart.py

```diff
@@ -21,159 +22,142 @@
+                    "lifetime": entry["lifetime_hours"],
+                    "lba_of_first_error": entry.get("lba"), # only included if there is an error
+                }
 
-            if test["lba_of_first_error"] == "-":
-                test["lba_of_first_error"] = None
+                if test["status_verbose"] == "Completed without error":
+                    test["status"] = "SUCCESS"
+                elif test["status_verbose"] == "Self-test routine in progress":
+                    test["status"] = "RUNNING"
+                elif test["status_verbose"] in ["Aborted by host", "Interrupted (host reset)"]:
+                    test["status"] = "ABORTED"
+                else:
+                    test["status"] = "FAILED"
 
-            tests.append(test)
+                tests.append(test)
 
         return tests
 
     # nvmeprint.cpp
-    if "Failing_LBA" in stdout:
-        got_header = False
-        for line in stdout.split("\n"):
-            if "Failing_LBA" in line:
-                got_header = True
-                continue
-
-            if not got_header:
-                continue
-
-            try:
-                status_verbose = line[23:56].strip()
-                if status_verbose == "Completed without error":
-                    status = "SUCCESS"
-                elif status_verbose.startswith("Aborted:"):
-                    status = "ABORTED"
-                else:
-                    status = "FAILED"
+    if "nvme_self_test_log" in data:
+        if "table" in data["nvme_self_test_log"]:
+            for index, entry in enumerate(data["nvme_self_test_log"]["table"]):
 
-                failing_lba = line[67:79].strip()
-                nsid = line[80:85].strip()
-                seg = line[86:89].strip()
-                sct = line[90:93]
-                code = line[94:98]
+                if lba := entry.get("lba"):
+                    lba = entry["lba"]["value"]
 
                 test = {
-                    "num": int(line[0:2].strip()),
-                    "description": line[5:22].strip(),
-                    "status": status,
-                    "status_verbose": status_verbose,
-                    "power_on_hours": int(line[57:66].strip()),
-                    "failing_lba": None if failing_lba == "-" else int(failing_lba),
-                    "nsid": None if nsid == "-" else nsid,
-                    "seg": None if seg == "-" else int(seg),
-                    "sct": sct,
-                    "code": code,
+                    "num": index,
+                    "description": entry["self_test_code"]["string"],
+                    "status_verbose": entry["self_test_result"]["string"],
+                    "power_on_hours": entry["power_on_hours"],
+                    "failing_lba": lba,
+                    "nsid": entry.get("nsid"),
+                    "seg": entry.get("segment"),
+                    "sct": entry.get("status_code_type") or 0x0,
```

Comment:
`sct` was a string before, but `0x0` is an integer

Maybe we can take advantage of this, define type classes and make this method return `list[AtaSelfTestLog | NvmeSelfTestLog | ScsiSelfTestLog]`?

Reply:
Unfortunately there's some implicit keys that are sometimes there or sometimes don't have a value. It'd be useful for the areas that consume those test logs, but I think it would get weirdly complicated to have type classes when the schema isn't consistent. 
It actually might be worth putting into another ticket though to look into further.

Reply:
The schema is consistent, the old code was returning the dictionaries with consistent sets of keys.

I propose to further formalize these dictionaries into dataclasses because I see that there is a confusion on which type `sct` field should have.